arci/src/clients/partial_joint_trajectory_client.rs

```diff
@@ -103,3 +103,410 @@ where
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::sync::{Arc, Mutex};
+
+    use assert_approx_eq::assert_approx_eq;
+
+    use super::*;
+
+    #[derive(Debug, Clone)]
+    struct DummyFull {
+        name: Vec<String>,
+        pos: Arc<Mutex<Vec<f64>>>,
+        last_trajectory: Arc<Mutex<Vec<TrajectoryPoint>>>,
+    }
+    impl JointTrajectoryClient for DummyFull {
+        fn joint_names(&self) -> Vec<String> {
+            self.name.clone()
+        }
+
+        fn current_joint_positions(&self) -> Result<Vec<f64>, Error> {
+            Ok(self.pos.lock().unwrap().clone())
+        }
+
+        fn send_joint_positions(
+            &self,
+            positions: Vec<f64>,
+            _duration: std::time::Duration,
+        ) -> Result<WaitFuture, Error> {
+            *self.pos.lock().unwrap() = positions;
+            Ok(WaitFuture::ready())
+        }
+
+        fn send_joint_trajectory(
+            &self,
+            full_trajectory: Vec<TrajectoryPoint>,
+        ) -> Result<WaitFuture, Error> {
+            if let Some(last_point) = full_trajectory.last() {
+                *self.pos.lock().unwrap() = last_point.positions.to_owned();
+            }
+            *self.last_trajectory.lock().unwrap() = full_trajectory;
+            Ok(WaitFuture::ready())
+        }
+    }
+
+    #[test]
+    fn test_partial_new() {
+        let client = DummyFull {
+            name: vec![
+                String::from("part1"),
+                String::from("high"),
+                String::from("part2"),
+            ],
+            pos: Arc::new(Mutex::new(vec![1.0_f64, 3.0_f64])),
+            last_trajectory: Arc::new(Mutex::new(Vec::new())),
+        };
+
+        // same joint pattern
+        let joint_names = vec![
+            String::from("part1"),
+            String::from("high"),
+            String::from("part2"),
+            String::from("low"),
+        ];
+        let partial = PartialJointTrajectoryClient::new(joint_names, client.clone());
+
+        assert_eq!(
+            format!("{:?}", partial.joint_names),
+            "[\"part1\", \"high\", \"part2\", \"low\"]"
+        );
+        assert_eq!(
+            format!("{:?}", client),
+            format!("{:?}", partial.shared_client)
+        );
+        assert_eq!(
+            format!("{:?}", partial.full_joint_names),
+            "[\"part1\", \"high\", \"part2\"]"
+        );
+
+        // few joint pattern
+        let joint_names = vec![String::from("part1"), String::from("high")];
+        let partial = PartialJointTrajectoryClient::new(joint_names, client.clone());
+
+        assert_eq!(
+            format!("{:?}", partial.joint_names),
+            "[\"part1\", \"high\"]"
+        );
+        assert_eq!(
+            format!("{:?}", client),
+            format!("{:?}", partial.shared_client)
+        );
+        assert_eq!(
+            format!("{:?}", partial.full_joint_names),
+            "[\"part1\", \"high\", \"part2\"]"
+        );
+    }
+
```

Comment:
What's the difference between `fn test_fn_copy_joint_position_for_from_joint()` and `fn test_fn_copy_joint_position_for_to_joint()`?